l3backend/l3backend-color.dtx

```diff
@@ -1248,6 +1253,401 @@
+%    \begin{macrocode}
+local function select(color_str)
+  set_attribute(color_attr,color_index(color_str))
+end
+ltxcolor.select = select
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{node_map, node_types,literals}
+%   We abstract the idea of different types of node where it makes sense:
+%   notice for example that in contrast to \pkg{luacolor} we extract some
+%   one-off information in the function that needs it. We also use a single
+%   local \enquote{map} for the node values.
+%    \begin{macrocode}
+local node_map = {
+    list         = 1 ,
+    list_leaders = 2 ,
+    list_disc    = 3 ,
+    color        = 4 ,
+    no_color     = 5
+  }
+local node_types = {
+    [id("hlist")]   = node_map.list      ,
+    [id("vlist")]   = node_map.list      ,
+    [id("rule")]    = node_map.color     ,
+    [id("glyph")]   = node_map.color     ,
+    [id("disc")]    = node_map.list_disc ,
+    [id("whatsit")] = {
+        [subtype("pdf_colorstack")] =
+          function(n) return n.stack == 0 and node_map.no_color or nil end ,
+        [subtype("special")]     = node_map.color ,
+        [subtype("pdf_literal")] = node_map.color ,
+        [subtype("pdf_save")]    = node_map.color ,
+        [subtype("pdf_restore")] = node_map.color ,
+      } ,
+    [id("glue")] =
+      function(n)
+        if n.subtype >= 100 then
+          if n.leader.id == id("rule") then
+            return node_map.color
+          else
+            return node_map.list_leaders
+          end
+        end
+      end
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{node_map, node_types,literals}
+%   We abstract the idea of different types of node where it makes sense:
+%   notice for example that in contrast to \pkg{luacolor} we extract some
+%   one-off information in the function that needs it. We also use a single
+%   local \enquote{map} for the node values.
+%    \begin{macrocode}
+local node_map = {
+    list         = 1 ,
+    list_leaders = 2 ,
+    list_disc    = 3 ,
+    color        = 4 ,
+    no_color     = 5
+  }
+local node_types = {
+    [id("hlist")]   = node_map.list      ,
+    [id("vlist")]   = node_map.list      ,
+    [id("rule")]    = node_map.color     ,
+    [id("glyph")]   = node_map.color     ,
+    [id("disc")]    = node_map.list_disc ,
+    [id("whatsit")] = {
+        [subtype("pdf_colorstack")] =
+          function(n) return n.stack == 0 and node_map.no_color or nil end ,
+        [subtype("special")]     = node_map.color ,
+        [subtype("pdf_literal")] = node_map.color ,
+        [subtype("pdf_save")]    = node_map.color ,
+        [subtype("pdf_restore")] = node_map.color
+      } ,
+    [id("glue")] =
+      function(n)
+        if n.subtype >= 100 then
+          if n.leader.id == id("rule") then
+            return node_map.color
+          else
+            return node_map.list_leaders
+          end
+        end
+      end
+  }
+%    \end{macrocode}
```

Comment:
This is duplicated.

Reply:
Fixed in 464b5be2fd5845dc1e0776b9cc74ffd33e93c007